### ISSUE TITLE
Procedural creation

### ISSUE THREAD

**Issue by MrTopom** [simlu/voxelshop#178]
As  I need it for my current project, I just initiate an issue related to this aspect.

I have to create different geometric objects of different sizes AND have some meta information extracted from the generation (Bounding box, surface perimeter...).

**Sample ideas:**

_Description_
`
box sx:10, sy:10, sz:20, id: base
meta-data src:base
    top-face
`

_VoxelShop Activity_
`
load-desc base.desc
save descs/base.vsd
export target-dir/base.dae
save-meta target-dir/base.meta.txt
`

would create a base.vsd file, the corresponding dae file and also the meta-information. Directly in the Unity directory.

**Comment by MrTopom** [comment]
The idea is to provide a set a simple command for usually thinks to do.

Sample 1: Provide some different roofs like here https://cdn-business.discourse.org/uploads/stonehearth/_optimized/1d7/9f2/1257bc33c63_690x417.png

Selecting the perimeter of your top face house you could then generate a roof from different roofs set.available.

Sample 2: Generate Castle or house with typical architecture aspect
- Change some voxels colors randomly (pseudo to get the same from one generation to the next one)
- Create windows/Doors with elements around
- Generate the building walls from base perimeter and extrusion
-...

Sample 3: Generate Some island/mountains/sea

Sample 4: In a model, define some water source and then add the voxels for the falling water.

**Comment by simlu** [comment]
So my initial idea is this:
You select the shape tool and a "variation". You then drag an area and it fills the area according to the variation. Variations are math formulas as described here: https://github.com/simlu/voxelshop/issues/43#issuecomment-258135187

**Comment by simlu** [comment]
This will be implemented as part of https://github.com/simlu/voxelshop/issues/43

Once that is done we will revisit what still needs to be done.

**Comment by simlu** [comment]
There are actually two tasks in here. One is the procedural creation and the other is exporting.

**Comment by MrTopom** [comment]
Procedural is different from tools.

- Tools allow interaction with the model in the scene by adding voxels.
- Behavior of tools is manage by an algorithm/variation who does the real voxel activity.
- Procedural generation use those behaviors to create things. 

So there is development to create new tools, based on new behaviors and on the other side create a procedural format/tiny language to be able to use those behaviors.

**Comment by simlu** [comment]
Yes, but the shape tool will use the "tiny programs". And a "procedural generation" is just a tool flavor with a different "tiny program" (algorithm).
